src/hardware/sblaster.cpp

```diff
@@ -1230,6 +1247,46 @@ static void flush_remainig_dma_transfer()
 		PIC_AddEvent(ProcessDMATransfer, delay, sb.dma.left);
 	}
 }
+
+static void per_tick_callback();
+
+static void per_frame_callback(uint32_t);
+
+static void add_next_frame_callback()
+{
+	assert(sblaster && sblaster->channel);
+	PIC_AddEvent(per_frame_callback, sblaster->channel->GetMillisPerFrame());
+}
+
+void CallbackTiming::Stop()
+{
+	if (timing != TimingType::None) {
+		(timing == TimingType::PerTick)
+		        ? TIMER_DelTickHandler(per_tick_callback)
+		        : PIC_RemoveEvents(per_frame_callback);
+		timing = TimingType::None;
+	}
+}
+
+void CallbackTiming::SetPerTick()
+{
+	if (timing != TimingType::PerTick) {
+		Stop();
+		frames_added_this_tick = 0;
+		TIMER_AddTickHandler(per_tick_callback);
+		timing = TimingType::PerTick;
+	}
+}
```

Comment:
I'm not really seeing it, but does this handle fractional "samples per tick" values? In general terms, that would be needed I think. But if it's just about shuffling code around, and the existing code only dealt with truncated integers, maybe it's fine for now.

Reply:
Yup; I've renamed this slightly to be clearer that we're just managing to two callback types (that differ on timing), as opposed to the timing itself.

I've added more comments for each callback (and noting here the handling of fractional frames that @weirddan455 added in the per-tick handler).

```c++
// This callback is run once per emulator tick (every 1ms), so it generates a
// batch of frames covering each 1ms time period. For example, if the Sound
// Blater's running at 8 Hz, then that's 8 frames per call. Many rates aren't
// evenly divisible by 1000 (For example, 22050 Hz is 22.05 frames/millisecond),
// so this function keeps track of exact fractional frames and uses rounding to
// ensure partial frames are accounted for and generated across N calls.
static void per_tick_callback()
{

// This callback is run exactly once per frame, so it only generates a single
// frame each call. This allows the emulator (and game) to run between each
// frame, which is more accurate because changes to the Sound Blaster are
// reflected in the next frame.
//
// This approach is more costly to emulate so we use it only when conditions
// demand it, and prefer the leaner per-tick approach otherwise. However,
// if/when the Sound Blaster is moved off the main emulator loop then we can
// (ideally) use this all of the time.
static void per_frame_callback(uint32_t)
```

Reply:
Very nicely explained, I really appreciate your effort to clearly document these things for the next guy 🎖️ 💯